opentracing_instrumentation/client_hooks/requests.py

```diff
@@ -25,44 +25,53 @@
 import logging
 import urllib.parse
 
+from opentracing.ext import tags
 from ..request_context import get_current_span
 from ..http_client import AbstractRequestWrapper
 from ..http_client import before_http_request
 from ..http_client import split_host_and_port
-from ._singleton import singleton
+from ._patcher import Patcher
 
 log = logging.getLogger(__name__)
 
 # Try to save the original entry points
 try:
-    import requests.sessions
     import requests.adapters
 except ImportError:  # pragma: no cover
     pass
 else:
     _HTTPAdapter_send = requests.adapters.HTTPAdapter.send
 
 
-@singleton
-def install_patches():
-    try:
-        import requests.sessions
-        import requests.adapters
-    except ImportError:  # pragma: no cover
-        return
-
-    def send_wrapper(self, request, **kwargs):
-        """Wraps HTTPAdapter.send"""
-
-        request_wrapper = RequestWrapper(request=request)
-        span = before_http_request(
-            request=request_wrapper,
-            current_span_extractor=get_current_span)
-        with span:
-            resp = _HTTPAdapter_send(self, request, **kwargs)
-            if hasattr(resp, 'status_code') and resp.status_code is not None:
-                span.set_tag('http.status_code', resp.status_code)
-        return resp
+class RequestsPatcher(Patcher):
+
+    applicable = '_HTTPAdapter_send' in globals()
```

Comment:
Q: in the base class this is declared as a readonly `@property`, and you don't seem to be overriding that function, but instead defining a class-level field. Could you elaborate?

Reply:
`@property` is used only for raising of `NotImplementedError` when `applicable` is not overridden by a child class.
There is not too much sense in overriding it with something like
```python
    @property
    def applicable(self):
        return '_HTTPAdapter_send' in globals()
```
since `_HTTPAdapter_send` should not magically appear in this module later, so I believe it's fine to calculate it only once on initialization of the class.

Reply:
+1